### PR TITLE
feature: code coverage support

### PR DESCRIPTION
**Pull request by Alizter**
[Coverage Status](https://coveralls.io/github/ocaml/dune?branch=main)

Apparently the badge will update when coverall detects it in our README.md.

Here are the coverage reports for Dune:

https://coveralls.io/github/ocaml/dune

### CODE REVIEW

File: .github/workflows/workflow.yml [hunk 1]
```diff
@@ -154,6 +154,42 @@ jobs:
...
+
+      # Install ocamlfind-secondary and ocaml-secondary-compiler, if needed
+      - run: opam install ./dune.opam --deps-only --with-test
+
+      - name: Install deps on Unix
+        run: |
+          opam install . --deps-only --with-test
+          opam exec -- make coverage-deps
+          opam exec -- make dev-deps
+          opam exec -- make coq-deps
```
Comment (rgrinberg) [inline .github/workflows/workflow.yml:186]: Can we exclude coq from coverage? Or have it exist in its own job.

File: dune-file [hunk 2]
```diff
@@ -45,6 +45,11 @@
  (copy dune-private-libs.opam.template chrome-trace.opam.template))
 
 (env
+ (coverage
+  (flags :standard -alert -unstable -w -58)
+  (env-vars
```
[the remark this answers is not in the document]
Comment (Alizter) [reply dune-file:50]: Without it the log is full of warnings.

File: .github/workflows/workflow.yml (hunk 1 above)
Comment (Alizter) [reply .github/workflows/workflow.yml:186]: Absolutely.

File: dune-file (hunk 2 above)
[the remark this answers is not in the document]
Comment (rgrinberg) [reply dune-file:50]: There has to be a bug somewhere then `(_ ..)` should set the env for everyone.

[the remark this answers is not in the document]
Comment (Alizter) [reply dune-file]: Huh seems to work now. I've removed the useless profile.

File: .github/workflows/workflow.yml [hunk 3]
```diff
@@ -154,6 +154,40 @@ jobs:
...
+          ocaml-compiler: ${{ matrix.ocaml-compiler }}
+          opam-depext: false
+
+      - name: Set git user
+        run: |
+          git config --global user.name github-actions[bot]
+          git config --global user.email github-actions[bot]@users.noreply.github.com
+
+      # Install ocamlfind-secondary and ocaml-secondary-compiler, if needed
+      - run: opam install ./dune.opam --deps-only --with-test
```
Comment (rgrinberg) [inline .github/workflows/workflow.yml:179]: `--with-test` can be safely removed here

File: Makefile [hunk 4]
```diff
@@ -77,9 +77,18 @@ melange:
 	opam pin add melange-compiler-libs https://github.com/melange-re/melange-compiler-libs.git#426463a77d0b70ecf0108c98e6a86d325cd01472
 	opam pin add melange https://github.com/melange-re/melange.git#685e546e290d317a884a4d48c7835467422c6426
 
+.PHONY: dev-deps
 dev-deps: melange
 	opam install -y $(TEST_DEPS)
 
+.PHONY: coq-deps
+coq-deps:
+	opam install -y coq.8.16.0 coq-native
```
Comment (rgrinberg) [inline Makefile:86]: do we still need this?

File: flake.nix [hunk 5]
```diff
@@ -42,6 +42,7 @@
         utop = "*";
         cinaps = "*";
         ocamlfind = "1.9.2";
+        bisect_ppx = "*";
```
Comment (rgrinberg) [inline flake.nix:45]: Do we need to run coverage locally? I'd remove this.

File: Makefile [hunk 6]
```diff
@@ -114,6 +119,10 @@ test-all: $(BIN)
 test-all-sans-melange: $(BIN)
 	$(BIN) build @runtest @runtest-js @runtest-coq
 
+test-coverage: $(BIN)
+	- $(BIN) build --instrument-with bisect_ppx --force @runtest
```
Comment (rgrinberg) [inline Makefile:123]: What does `-` do in front of a command?

Comment (Alizter) [reply Makefile:123]: It ignores the exit code allowing the command to fail.

Comment (Alizter) [reply Makefile:123]: We know that not all the tests will run when coverage is enabled, so we do a best case scenario (hence --force).

**Pull request merged**